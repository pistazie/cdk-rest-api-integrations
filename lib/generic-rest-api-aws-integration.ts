import * as cdk from '@aws-cdk/core';
import * as iam from '@aws-cdk/aws-iam';
import * as apigateway from '@aws-cdk/aws-apigateway';

export interface GenericRestApiAwsIntegrationProps {
  /**
   * The RestApi resource to which a method will be added and integrated with the AWS service
   */
  restApiResource: apigateway.Resource;

  /**
   * The status code returned by RestApi on a successful request
   *
   * @default "200"
   */
  successResponseStatusCode?: string;

  /**
   * The status code returned by RestApi on a failing request
   *
   * @default "503"
   */
  failureResponseStatusCode?: string;

}

export abstract class GenericRestApiAwsIntegration extends cdk.Construct {
  /** @returns RestApi::Method to which the AwsIntegration was integrated to */
  protected readonly restApiMethod: apigateway.Method;

  protected abstract configureAwsService(id:string, props: GenericRestApiAwsIntegrationProps) : void;
  protected abstract init(id:string, props: GenericRestApiAwsIntegrationProps) : void;

  // configured in init method
  protected awsService: string
  protected integrationHttpVerb = 'POST'
  protected integrationPath = "/"
  protected requestTemplates: { [contentType: string]: string; } = {}
  protected successResponseTemplates: { [contentType: string]: string; } = {}
  protected failureResponseTemplates: { [contentType: string]: string; } = {}

  public readonly integrationRole: iam.IRole

  protected constructor(scope: cdk.Construct, id: string, props: GenericRestApiAwsIntegrationProps) {
    super(scope, id);

    this.init(id,props)
    this.integrationRole = this.addIntegrationRole(id)
    const awsIntegration = this.addIntegration(props)

    this.restApiMethod = props.restApiResource.addMethod(this.integrationHttpVerb, awsIntegration, {
      methodResponses: [{
        statusCode: props.failureResponseStatusCode || "503",
        responseModels: {
          "application/json": apigateway.Model.EMPTY_MODEL
        }
      }, {
        statusCode: props.successResponseStatusCode || "200",
        responseModels: {
          "application/json": apigateway.Model.EMPTY_MODEL
        }
      }
      ]
    })

    this.configureAwsService(id,props)
  }

  private addIntegration(props: GenericRestApiAwsIntegrationProps) {

    const awsIntegration = new apigateway.AwsIntegration({
      service: this.awsService,
      path: this.integrationPath,
      options: {
        credentialsRole: this.integrationRole,
        passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
        requestParameters: {
          "integration.request.header.Content-Type": "'application/x-www-form-urlencoded'"
        },
        // useful for working with these templates: http://mapping-template-checker.toqoz.net/
        // variable reference:  https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html
        requestTemplates: this.requestTemplates,
        integrationResponses: [
          {
            statusCode: props.failureResponseStatusCode || "503",
            responseTemplates: this.failureResponseTemplates,
            // https://docs.aws.amazon.com/apigateway/api-reference/resource/integration-response/#selectionPattern
            selectionPattern: "" // --> this is the default response
          },
          {
            selectionPattern: "200",
            statusCode: props.successResponseStatusCode || "200",
            responseTemplates: this.successResponseTemplates
          }
        ]
      }
    })
    return awsIntegration
  }

  private addIntegrationRole(id: string) {
    return new iam.Role(this, `${id}-${this.awsService}-integration-role`, {
      assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com')
    })
  }
}
