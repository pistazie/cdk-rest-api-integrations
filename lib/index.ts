export * from './sns';
export * from './sqs';
export * from './kinesis';