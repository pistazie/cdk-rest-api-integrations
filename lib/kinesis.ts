import * as kinesis from '@aws-cdk/aws-kinesis';
import * as cdk from '@aws-cdk/core';
import {GenericRestApiAwsIntegration, GenericRestApiAwsIntegrationProps} from "./generic-rest-api-aws-integration"

/**
 * Integrates a Kinesis Stream with ApiGateway (RestApi) by publishing Queue messages on http request
 */
export interface KinesisRestApiIntegrationProps extends GenericRestApiAwsIntegrationProps {
    /**
     * The target kinesis Queue to which messages will be sent by the RestApi Integration
     */
    stream: kinesis.Stream;
}

export class KinesisRestApiIntegration extends GenericRestApiAwsIntegration {

    constructor(scope: cdk.Construct, id: string, props: KinesisRestApiIntegrationProps) {
        super(scope, id, props);
    }

    init(id: string, props: KinesisRestApiIntegrationProps) : void {

        this.requestTemplates["application/json"] = "{\"StreamName\":\""+ props.stream.streamName +"\",\"Data\":\"$util.base64Encode($input.body)\",\"PartitionKey\":\"$context.requestId\"}"
        this.successResponseTemplates["application/json"] = "{\"status\":\"message received\", \"messageId\": $input.json('SequenceNumber')}"
        this.failureResponseTemplates["application/json"] = "{\"status\":\"failed to process message\")}"
        this.integrationPath = undefined
        this.integrationAction = "PutRecord"
        this.awsService = "kinesis"
    }

    configureAwsService(id: string, props: KinesisRestApiIntegrationProps) : void {
        props.stream.grantWrite(this.integrationRole)
    }
}
