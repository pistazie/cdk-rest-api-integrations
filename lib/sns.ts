import * as sns from '@aws-cdk/aws-sns';
import * as cdk from '@aws-cdk/core';
import {GenericRestApiAwsIntegration, GenericRestApiAwsIntegrationProps} from "./generic-rest-api-aws-integration"

/**
 * Integrates an SNS Topic with ApiGateway(RestApi) by publishing Topic messages on http request
 */
export interface SnsApiAwsIntegrationProps extends GenericRestApiAwsIntegrationProps {
    /**
     * The target SNS Topic to which messages will be published by the RestApi Integration
     */
    topic: sns.Topic;
}

export class SnsApiAwsIntegration extends GenericRestApiAwsIntegration {

    constructor(scope: cdk.Construct, id: string, props: SnsApiAwsIntegrationProps) {
        super(scope, id, props);
    }

    init(id: string, props: SnsApiAwsIntegrationProps) : void {
        this.requestTemplates["application/json"] = "Action=Publish&TopicArn=$util.urlEncode('" + props.topic.topicArn + "')&Message=$util.urlEncode($input.body)"
        this.successResponseTemplates["application/json"] = "{\"status\":\"message received\", \"messageId\": $input.json('PublishResponse.PublishResult.MessageId')}"
        this.failureResponseTemplates["application/json"] = "{\"status\":\"failed to process message\")}"
        this.integrationPath="/"
        this.awsService = "sns"
    }

    configureAwsService(id: string, props: SnsApiAwsIntegrationProps): void {
        props.topic.grantPublish(this.integrationRole)
    }
}
