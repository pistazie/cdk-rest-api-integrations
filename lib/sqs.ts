import * as sqs from '@aws-cdk/aws-sqs';
import * as cdk from '@aws-cdk/core';
import {GenericRestApiAwsIntegration, GenericRestApiAwsIntegrationProps} from "./generic-rest-api-aws-integration"

/**
 * Integrates an SQS Queue with ApiGateway (RestApi) by publishing Queue messages on http request
 */
export interface SqsRestApiIntegrationProps extends GenericRestApiAwsIntegrationProps {
    /**
     * The target SQS Queue to which messages will be sent by the RestApi Integration
     */
    queue: sqs.IQueue;

    /**
     *  Integration request mapping templates
     *
     *  reference: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html
     *
     *  @default {"application/json": "Action=SendMessage&MessageBody=$util.urlEncode(\"$input.body\")"}
     */
    requestTemplates?: { [contentType: string]: string; };
}

export class SqsRestApiIntegration extends GenericRestApiAwsIntegration {

    constructor(scope: cdk.Construct, id: string, props: SqsRestApiIntegrationProps) {
        super(scope, id, props);
    }

    init(id: string, props: SqsRestApiIntegrationProps) : void {

        this.requestTemplates = props.requestTemplates
            || {"application/json": "Action=SendMessage&MessageBody=$util.urlEncode(\"$input.body\")"}
        this.successResponseTemplates["application/json"] = "{\"status\":\"message received\", \"messageId\": $input.json('SendMessageResponse.SendMessageResult.MessageId')}"
        this.failureResponseTemplates["application/json"] = "{\"status\":\"failed to process message\")}"
        this.integrationPath = cdk.Aws.ACCOUNT_ID + "/" + props.queue.queueName
        this.awsService = "sqs"
    }

    configureAwsService(id: string, props: SqsRestApiIntegrationProps) : void {
        props.queue.grantSendMessages(this.integrationRole)
    }
}
