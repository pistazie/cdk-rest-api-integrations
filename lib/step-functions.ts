import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as cdk from '@aws-cdk/core';
import {GenericRestApiAwsIntegration, GenericRestApiAwsIntegrationProps} from "./generic-rest-api-aws-integration"

/**
 * Integrates a Step Function with ApiGateway(RestApi) by starting an Execution on http request
 */
export interface StepFunctionsRestApiIntegrationProps extends GenericRestApiAwsIntegrationProps {
    /**
     * The target Step Functions StateMachine which will be executed by the RestApi Integration
     */
    stateMachine: sfn.StateMachine;

    /**
     *  Integration request mapping templates
     *
     *  reference: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html
     *
     *  @default {"application/json": "Action=Publish&TopicArn=$util.urlEncode('" + props.stateMachine.topicArn + "')&Message=$util.urlEncode($input.body)"}
     */
    requestTemplates?: { [contentType: string]: string; };

}

export class StepFunctionsRestApiIntegration extends GenericRestApiAwsIntegration {

    constructor(scope: cdk.Construct, id: string, props: StepFunctionsRestApiIntegrationProps) {
        super(scope, id, props);
    }

    init(id: string, props: StepFunctionsRestApiIntegrationProps) : void {
        this.requestTemplates = props.requestTemplates ||  { "application/json": "{\"stateMachineArn\":\""+ props.stateMachine.stateMachineArn +"\",\"input\":\"$util.escapeJavaScript($input.body)\"}"}
        this.successResponseTemplates["application/json"] = "{\"status\":\"execution started\", \"executionId\": $input.json('$.executionArn').split(':')[7]}"
        this.failureResponseTemplates["application/json"] = "{\"status\":\"failed to start the execution\")}"
        this.integrationPath=undefined
        this.integrationAction = "StartExecution"
        this.awsService = "states"
    }

    configureAwsService(id: string, props: StepFunctionsRestApiIntegrationProps): void {
        props.stateMachine.grantStartExecution(this.integrationRole)
    }
}
