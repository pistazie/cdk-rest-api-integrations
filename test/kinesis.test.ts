import * as cdk from '@aws-cdk/core';
import * as apigateway from '@aws-cdk/aws-apigateway';
import * as kinesis from '@aws-cdk/aws-kinesis';

import {SynthUtils} from '@aws-cdk/assert';
import {KinesisRestApiIntegration} from "../lib"

test('snapshot test: Kinesis integration as expected', () => {

    const {stack, stream, restApiResource} = givenStackRestApiResourceAndQueue()

    new KinesisRestApiIntegration(stack, 'testIntegration', {
        restApiResource: restApiResource,
        stream: stream
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

function givenStackRestApiResourceAndQueue() {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, "TestStack");
    const restApi = new apigateway.RestApi(stack, 'testRestApi')
    const restApiResource = restApi.root.addResource("testResource")
    const stream = new kinesis.Stream(stack, "testStream")
    return {stack, restApi, stream, restApiResource}
}
