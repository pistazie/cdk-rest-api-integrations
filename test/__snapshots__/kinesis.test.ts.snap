// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: Kinesis integration as expected 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Outputs": Object {
    "testRestApiEndpointA2F50208": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "testRestApi5A6FA7A3",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "testRestApiDeploymentStageprod5AB21DDE",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Resources": Object {
    "testIntegrationtestIntegrationkinesisintegrationrole6E377ED1": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testIntegrationtestIntegrationkinesisintegrationroleDefaultPolicyFA24CC1B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:ListShards",
                "kinesis:PutRecord",
                "kinesis:PutRecords",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testStream8BCA7523",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testIntegrationtestIntegrationkinesisintegrationroleDefaultPolicyFA24CC1B",
        "Roles": Array [
          Object {
            "Ref": "testIntegrationtestIntegrationkinesisintegrationrole6E377ED1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testRestApi5A6FA7A3": Object {
      "Properties": Object {
        "Name": "testRestApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "testRestApiAccount78640F48": Object {
      "DependsOn": Array [
        "testRestApi5A6FA7A3",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "testRestApiCloudWatchRole995562DC",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "testRestApiCloudWatchRole995562DC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testRestApiDeployment1ED9BE2E3f4f7fb0086c4d79c7c3d1d66fa2cce3": Object {
      "DependsOn": Array [
        "testRestApitestResourcePOST4F369133",
        "testRestApitestResourceD4EBD8FD",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "testRestApi5A6FA7A3",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "testRestApiDeploymentStageprod5AB21DDE": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "testRestApiDeployment1ED9BE2E3f4f7fb0086c4d79c7c3d1d66fa2cce3",
        },
        "RestApiId": Object {
          "Ref": "testRestApi5A6FA7A3",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "testRestApitestResourceD4EBD8FD": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "testRestApi5A6FA7A3",
            "RootResourceId",
          ],
        },
        "PathPart": "testResource",
        "RestApiId": Object {
          "Ref": "testRestApi5A6FA7A3",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "testRestApitestResourcePOST4F369133": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "testIntegrationtestIntegrationkinesisintegrationrole6E377ED1",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "ResponseTemplates": Object {
                "application/json": "{\\"status\\":\\"failed to process message\\")}",
              },
              "SelectionPattern": "",
              "StatusCode": "503",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "{\\"status\\":\\"message received\\", \\"messageId\\": $input.json('SequenceNumber')}",
              },
              "SelectionPattern": "200",
              "StatusCode": "200",
            },
          ],
          "PassthroughBehavior": "NEVER",
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/x-www-form-urlencoded'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{\\"StreamName\\":\\"",
                  Object {
                    "Ref": "testStream8BCA7523",
                  },
                  "\\",\\"Data\\":\\"$util.base64Encode($input.body)\\",\\"PartitionKey\\":\\"$context.requestId\\"}",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":kinesis:action/PutRecord",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "ResponseModels": Object {
              "application/json": "Empty",
            },
            "StatusCode": "503",
          },
          Object {
            "ResponseModels": Object {
              "application/json": "Empty",
            },
            "StatusCode": "200",
          },
        ],
        "ResourceId": Object {
          "Ref": "testRestApitestResourceD4EBD8FD",
        },
        "RestApiId": Object {
          "Ref": "testRestApi5A6FA7A3",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "testStream8BCA7523": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
